File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/scala/src/main/scala/proven/SafeNetwork.scala
// SPDX-License-Identifier: PMPL-1.0

package proven

import java.net.{InetAddress, URI}
import scala.util.Try

/**
 * Classification of IP addresses.
 */
sealed trait IpClassification
object IpClassification {
  case object Loopback extends IpClassification
  case object Private extends IpClassification
  case object Reserved extends IpClassification
  case object Public extends IpClassification
  case object Invalid extends IpClassification
}

/**
 * Parsed IPv4 address.
 */
case class IPv4Address(a: Int, b: Int, c: Int, d: Int) {

  /** Convert to integer representation */
  def toInt: Long = ((a.toLong & 0xFF) << 24) | ((b & 0xFF) << 16) | ((c & 0xFF) << 8) | (d & 0xFF)

  /** Check if this is a loopback address (127.0.0.0/8) */
  def isLoopback: Boolean = a == 127

  /** Check if this is a private address (RFC 1918) */
  def isPrivate: Boolean = {
    // 10.0.0.0/8
    if (a == 10) return true

    // 172.16.0.0/12
    if (a == 172 && b >= 16 && b <= 31) return true

    // 192.168.0.0/16
    if (a == 192 && b == 168) return true

    false
  }

  /** Check if this is a reserved address */
  def isReserved: Boolean = {
    // 0.0.0.0/8
    if (a == 0) return true

    // 100.64.0.0/10 - CGNAT
    if (a == 100 && b >= 64 && b <= 127) return true

    // 169.254.0.0/16 - Link-local
    if (a == 169 && b == 254) return true

    // 192.0.0.0/24 - IETF
    if (a == 192 && b == 0 && c == 0) return true

    // 192.0.2.0/24 - TEST-NET-1
    if (a == 192 && b == 0 && c == 2) return true

    // 198.51.100.0/24 - TEST-NET-2
    if (a == 198 && b == 51 && c == 100) return true

    // 203.0.113.0/24 - TEST-NET-3
    if (a == 203 && b == 0 && c == 113) return true

    // 224.0.0.0/4 - Multicast
    if (a >= 224 && a <= 239) return true

    // 240.0.0.0/4 - Reserved
    if (a >= 240) return true

    false
  }

  /** Check if this is a public address */
  def isPublic: Boolean = !isLoopback && !isPrivate && !isReserved

  /** Get classification of this address */
  def classification: IpClassification = {
    if (isLoopback) IpClassification.Loopback
    else if (isPrivate) IpClassification.Private
    else if (isReserved) IpClassification.Reserved
    else IpClassification.Public
  }

  /** Check if address is in a CIDR range */
  def isInRange(network: IPv4Address, prefixLength: Int): Boolean = {
    if (prefixLength < 0 || prefixLength > 32) return false

    val mask = if (prefixLength == 0) 0L else (0xFFFFFFFFL << (32 - prefixLength))
    (toInt & mask) == (network.toInt & mask)
  }

  override def toString: String = s"$a.$b.$c.$d"
}

object IPv4Address {
  private val OctetPattern = """^[0-9]+$""".r

  /**
   * Parse an IPv4 address string.
   */
  def parse(address: String): Option[IPv4Address] = {
    val parts = address.split("\\.")
    if (parts.length != 4) return None

    val octets = parts.flatMap { part =>
      // Check for empty parts
      if (part.isEmpty) return None

      // Check for leading zeros (invalid in strict parsing)
      if (part.length > 1 && part.startsWith("0")) return None

      // Check for non-digit characters
      if (OctetPattern.findFirstIn(part).isEmpty) return None

      // Parse and validate range
      Try(part.toInt).toOption.filter(o => o >= 0 && o <= 255)
    }

    if (octets.length != 4) None
    else Some(IPv4Address(octets(0), octets(1), octets(2), octets(3)))
  }
}

/**
 * Safe network validation and operations.
 */
object SafeNetwork {

  /**
   * Check if string is a valid IPv4 address.
   */
  def isValidIPv4(address: String): Boolean = IPv4Address.parse(address).isDefined

  /**
   * Check if string is a valid IPv6 address.
   */
  def isValidIPv6(address: String): Boolean = {
    Try {
      val inet = InetAddress.getByName(address)
      inet.getAddress.length == 16 && address.contains(":")
    }.getOrElse(false)
  }

  /**
   * Check if string is any valid IP address.
   */
  def isValidIP(address: String): Boolean = isValidIPv4(address) || isValidIPv6(address)

  /**
   * Parse an IPv4 address.
   */
  def parseIPv4(address: String): Option[IPv4Address] = IPv4Address.parse(address)

  /**
   * Classify an IPv4 address.
   */
  def classifyIPv4(address: String): IpClassification = {
    IPv4Address.parse(address) match {
      case Some(ip) => ip.classification
      case None => IpClassification.Invalid
    }
  }

  /**
   * Check if port number is valid (1-65535).
   */
  def isValidPort(port: Int): Boolean = port >= 1 && port <= 65535

  /**
   * Check if port is privileged (< 1024).
   */
  def isPrivilegedPort(port: Int): Boolean = port >= 1 && port < 1024

  /**
   * Check if host:port string is valid.
   */
  def isValidHostPort(hostPort: String): Boolean = {
    val lastColon = hostPort.lastIndexOf(':')
    if (lastColon == -1) return false

    val host = hostPort.substring(0, lastColon)
    val portStr = hostPort.substring(lastColon + 1)

    val port = Try(portStr.toInt).toOption
    if (port.isEmpty || !isValidPort(port.get)) return false

    isValidIP(host) || isValidHostname(host)
  }

  /**
   * Check if string is a valid hostname.
   */
  def isValidHostname(hostname: String): Boolean = {
    if (hostname.isEmpty || hostname.length > 253) return false

    val labels = hostname.split("\\.")
    labels.forall { label =>
      label.nonEmpty && label.length <= 63 &&
        """^[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?$""".r.findFirstIn(label).isDefined
    }
  }

  /**
   * Check if string is a valid URL.
   */
  def isValidUrl(url: String): Boolean = {
    Try {
      val uri = new URI(url)
      uri.getScheme != null && (uri.getScheme == "http" || uri.getScheme == "https")
    }.getOrElse(false)
  }

  /**
   * Parse a URL safely.
   */
  def parseUrl(url: String): Option[URI] = {
    Try(new URI(url)).toOption.filter(_.getScheme != null)
  }

  /**
   * Check if URL host is a private IP (SSRF protection).
   */
  def isPrivateUrl(url: String): Boolean = {
    parseUrl(url).exists { uri =>
      val host = uri.getHost
      if (host == null) return false

      // Check for localhost
      if (host == "localhost" || host == "127.0.0.1" || host == "::1") {
        return true
      }

      // Check IPv4
      IPv4Address.parse(host).exists { ip =>
        ip.isPrivate || ip.isLoopback || ip.isReserved
      }
    }
  }

  /**
   * Format IPv4 from octets.
   */
  def formatIPv4(a: Int, b: Int, c: Int, d: Int): String = s"$a.$b.$c.$d"
}

File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/clojure/src/proven/safe_network.clj
;; SPDX-License-Identifier: PMPL-1.0
;; SPDX-FileCopyrightText: 2025 Hyperpolymath

(ns proven.safe-network
  "Safe network validation and operations."
  (:require [clojure.string :as str])
  (:import [java.net InetAddress URI]))

(defrecord IPv4Address [a b c d])

(defn parse-ipv4
  "Parse an IPv4 address string. Returns IPv4Address or nil."
  [address]
  (let [parts (str/split address #"\.")]
    (when (= 4 (count parts))
      (let [octets (map (fn [part]
                          (when (and (not (str/blank? part))
                                     (re-matches #"^[0-9]+$" part)
                                     (or (= part "0")
                                         (not (str/starts-with? part "0"))))
                            (try
                              (let [n (Long/parseLong part)]
                                (when (<= 0 n 255) n))
                              (catch NumberFormatException _ nil))))
                        parts)]
        (when (every? some? octets)
          (->IPv4Address (nth octets 0) (nth octets 1) (nth octets 2) (nth octets 3)))))))

(defn ipv4->int
  "Convert IPv4Address to integer representation."
  [{:keys [a b c d]}]
  (bit-or (bit-shift-left (bit-and a 0xFF) 24)
          (bit-shift-left (bit-and b 0xFF) 16)
          (bit-shift-left (bit-and c 0xFF) 8)
          (bit-and d 0xFF)))

(defn loopback?
  "Check if IPv4 is a loopback address (127.0.0.0/8)."
  [{:keys [a]}]
  (= a 127))

(defn private?
  "Check if IPv4 is a private address (RFC 1918)."
  [{:keys [a b]}]
  (or (= a 10)
      (and (= a 172) (<= 16 b 31))
      (and (= a 192) (= b 168))))

(defn reserved?
  "Check if IPv4 is a reserved address."
  [{:keys [a b c]}]
  (or (= a 0)
      (and (= a 100) (<= 64 b 127))
      (and (= a 169) (= b 254))
      (and (= a 192) (= b 0) (= c 0))
      (and (= a 192) (= b 0) (= c 2))
      (and (= a 198) (= b 51) (= c 100))
      (and (= a 203) (= b 0) (= c 113))
      (<= 224 a 239)
      (>= a 240)))

(defn public?
  "Check if IPv4 is a public address."
  [ip]
  (not (or (loopback? ip) (private? ip) (reserved? ip))))

(defn classify
  "Get classification of IPv4 address."
  [ip]
  (cond
    (loopback? ip) :loopback
    (private? ip) :private
    (reserved? ip) :reserved
    :else :public))

(defn in-range?
  "Check if IPv4 address is in a CIDR range."
  [ip network prefix-length]
  (when (and (>= prefix-length 0) (<= prefix-length 32))
    (let [mask (if (= prefix-length 0)
                 0
                 (bit-shift-left 0xFFFFFFFF (- 32 prefix-length)))]
      (= (bit-and (ipv4->int ip) mask)
         (bit-and (ipv4->int network) mask)))))

(defn valid-ipv4?
  "Check if string is a valid IPv4 address."
  [address]
  (some? (parse-ipv4 address)))

(defn valid-ipv6?
  "Check if string is a valid IPv6 address."
  [address]
  (try
    (let [inet (InetAddress/getByName address)]
      (and (= 16 (count (.getAddress inet)))
           (str/includes? address ":")))
    (catch Exception _ false)))

(defn valid-ip?
  "Check if string is any valid IP address."
  [address]
  (or (valid-ipv4? address) (valid-ipv6? address)))

(defn classify-ipv4
  "Classify an IPv4 address string."
  [address]
  (if-let [ip (parse-ipv4 address)]
    (classify ip)
    :invalid))

(defn valid-port?
  "Check if port number is valid (1-65535)."
  [port]
  (<= 1 port 65535))

(defn privileged-port?
  "Check if port is privileged (< 1024)."
  [port]
  (and (>= port 1) (< port 1024)))

(defn valid-hostname?
  "Check if string is a valid hostname."
  [hostname]
  (when (and (not (str/blank? hostname))
             (<= (count hostname) 253))
    (let [labels (str/split hostname #"\.")]
      (every? (fn [label]
                (and (not (str/blank? label))
                     (<= (count label) 63)
                     (re-matches #"^[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?$" label)))
              labels))))

(defn valid-host-port?
  "Check if host:port string is valid."
  [host-port]
  (let [last-colon (str/last-index-of host-port ":")]
    (when last-colon
      (let [host (subs host-port 0 last-colon)
            port-str (subs host-port (inc last-colon))]
        (when-let [port (try (Long/parseLong port-str) (catch Exception _ nil))]
          (and (valid-port? port)
               (or (valid-ip? host) (valid-hostname? host))))))))

(defn valid-url?
  "Check if string is a valid URL."
  [url]
  (try
    (let [uri (URI. url)]
      (and (.getScheme uri)
           (#{"http" "https"} (.getScheme uri))))
    (catch Exception _ false)))

(defn parse-url
  "Parse a URL safely. Returns URI or nil."
  [url]
  (try
    (let [uri (URI. url)]
      (when (.getScheme uri) uri))
    (catch Exception _ nil)))

(defn private-url?
  "Check if URL host is a private IP (SSRF protection)."
  [url]
  (when-let [uri (parse-url url)]
    (let [host (.getHost uri)]
      (or (= host "localhost")
          (= host "127.0.0.1")
          (= host "::1")
          (when-let [ip (parse-ipv4 host)]
            (or (private? ip) (loopback? ip) (reserved? ip)))))))

(defn format-ipv4
  "Format IPv4 from octets."
  [a b c d]
  (str a "." b "." c "." d))
